File: bonus/cheapest_flights_within_k_stops.js
/**
 * @param {number} n
 * @param {number[][]} flights
 * @param {number} src
 * @param {number} dst
 * @param {number} k
 * @return {number}
 */
var findCheapestPriceDS1 = function(n, flights, src, dst, k)
{
    let adj = new Map();
    for (const flight of flights)
    {
        if (adj.has(flight[0]))
        {
            adj.get(flight[0]).push([flight[1], flight[2]]);
        }
        else
        {
            adj.set(flight[0], [[flight[1], flight[2]]]);
        }
    }

    let costs = new Array(n).fill(Number.MAX_SAFE_INTEGER);

    let city_cost = [[src, 0]];

    let stops = 0;
    while (stops <= k && city_cost.length)
    {
        let level_size = city_cost.length;
        let next_city_cost = [];

        for (let idx = 0; idx < level_size; idx++)
        {
            const [curr_city, cost_to_curr_city] = city_cost[idx];

            if (!adj.has(curr_city))
            {
                continue;
            }

            for (const [neighbor, price] of adj.get(curr_city))
            {
                if (cost_to_curr_city + price >= costs[neighbor])
                {
                    continue;
                }

                costs[neighbor] = cost_to_curr_city + price;
                next_city_cost.push([neighbor, costs[neighbor]]);
            }
        }

        city_cost = next_city_cost;
        stops++;
    }

    return costs[dst] === Number.MAX_SAFE_INTEGER ? -1 : costs[dst];
};

/**
 * @param {number} n
 * @param {number[][]} flights
 * @param {number} src
 * @param {number} dst
 * @param {number} k
 * @return {number}
 */
var findCheapestPriceDS2 = function(n, flights, src, dst, k)
{
    // Costs from src to all other cities
    let costs = new Array(n).fill(Number.MAX_SAFE_INTEGER);
    costs[src] = 0;

    // Run only k + 1 times since want minimum cost in k stops
    for (let num_flights = 0; num_flights <= k; num_flights++)
    {
        // Create copy of costs array
        let temp = [...costs];

        for (const [from, to, price] of flights)
        {
            if (costs[from] != Number.MAX_SAFE_INTEGER)
            {
                temp[to] = Math.min(temp[to], costs[from] + price);
            }
        }

        // Copy temp array into costs
        costs = temp;
    }

    return costs[dst] == Number.MAX_SAFE_INTEGER ? -1 : costs[dst];
};

File: bonus/cheapest_flights_within_k_stops.py
from collections import defaultdict
from collections import deque


def findCheapestPriceDS1(
    n: int, flights: list[list[int]], src: int, dst: int, k: int
) -> int:
    adj = defaultdict(list)
    for flight in flights:
        adj[flight[0]].append([flight[1], flight[2]])

    costs = [float('inf')] * n

    city_cost = deque([(src, 0)])

    stops = 0
    while stops <= k and city_cost:
        level_size = len(city_cost)

        for _ in range(level_size, 0, -1):
            curr_city, cost_to_curr_city = city_cost.popleft()

            for neighbor, price in adj[curr_city]:
                if cost_to_curr_city + price >= costs[neighbor]:
                    continue

                costs[neighbor] = cost_to_curr_city + price
                city_cost.append((neighbor, costs[neighbor]))

        stops += 1

    return -1 if costs[dst] == float('inf') else costs[dst]


def findCheapestPriceDS2(
    n: int, flights: list[list[int]], src: int, dst: int, k: int
) -> int:
    # Costs from src to all other cities
    costs = [float('inf')] * n
    costs[src] = 0

    # Run only k + 1 times since want minimum cost in k stops
    for _ in range(k + 1):
        # Create copy of costs array
        temp = costs.copy()

        for flight in flights:
            if costs[flight[0]] != float('inf'):
                temp[flight[1]] = min(temp[flight[1]],
                                      costs[flight[0]] + flight[2])

        costs = temp

    return -1 if costs[dst] == float('inf') else costs[dst]
